### PR TITLE
Thumbnails for url identifiers

### PR DESCRIPTION
**Pull request by kirkkwang**
# Story

[🐛 Fix URI identifiers and URI's without pdf](https://github.com/scientist-softserv/west-virginia-university/commit/c38d200f72a265e7f6090248360528a427e7579d) 

This commit will modify the lookup logic for the record based on the id
instead of the identifier.  The problem was because now that we are
accepting url identifiers, the directories and filenames that are based
of off identifiers are not valid for url identifiers because they
contain invalid characters.  Since ids are already sanitized for special
characters, we will use those instead to make valid directory and
file names.

Also in this commit we move the MIME type check logic into the
`GenerateThumbsJob` so we can determine what type of file we are
downloading prior to picking a thumbnail generator.  Checking the MIME
type is more durable than checking the string of the URI.

[🐛 Move database calls out of the views](https://github.com/scientist-softserv/west-virginia-university/commit/cef75968608a1e8f8ac080e6639ea184d58df917) 

This commit will fix a bug where the id was saved as the identifier.
Though not truly a bug but more of a misnomer, it however did affect
URI identifiers because identifiers were not the same as the id in this
case.  This commit will also move the database calls from the views to
the controller making it adhere closer to the MVC pattern.

### CODE REVIEW

File: hydra/app/controllers/catalog_controller.rb [hunk 1]
```diff
@@ -195,7 +195,11 @@ class CatalogController < ApplicationController
   end
 
   def index
-    return super if params[:range].nil?
+    if params[:range].nil?
```
Comment (jeremyf) [inline hydra/app/controllers/catalog_controller.rb]: The previous method returned the results of `super`; which may or may not be material.  In the case of a controller action it is like immaterial.  However, it technically changes the interface.

Consider the following for logic inside of the `if` statement:

```
returning_value = super
fetched_image_files
return returning_value
```

File: hydra/app/controllers/catalog_controller.rb [hunk 2]
```diff
@@ -207,16 +211,20 @@ def index
     end
 
     super
+    fetch_image_files
```
Comment (jeremyf) [inline hydra/app/controllers/catalog_controller.rb]: See above.

File: hydra/app/controllers/catalog_controller.rb [hunk 3]
```diff
@@ -207,16 +211,20 @@ def index
...
-  #   super
-  #   @metadata = []
-  #   # loop over each field and add it to the metadata
-  #   @document.each_pair do |k,v|
-  #     @metadata << [k, v] unless v.present?
-  #   end
-  # end
+  def show
+    super
+    # @metadata = []
```
Comment (jeremyf) [inline hydra/app/controllers/catalog_controller.rb]: What's the purpose of carrying these code comments forward?  Is that the original `super` method?

Also see above about method signature (e.g. returning value)

File: hydra/app/controllers/catalog_controller.rb [hunk 4]
```diff
@@ -253,4 +261,14 @@ def policies
   def contributingcollections
     render "contributingcollections.html.erb"
   end
+
+  private
+
+    # Fetches the image files for each document
+    def fetch_image_files
+      @image_files = @document_list.each_with_object({}) do |document, hash|
+        acda_record = Acda.where(id: document.id).first
```
Comment (jeremyf) [inline hydra/app/controllers/catalog_controller.rb]: This is a very non-performant query.  We're hitting SOLR one additional time per item in the document list.  We should instead be writing a more customized query that takes the IDs from the document_list and passes those as a single query to Solr.  Also we only need the `id` and `image_file` attributes from the sole document.

[This logic provides some guidance](https://github.com/samvera-labs/bulkrax/blob/657c6feaae8d81fe9487dbd16dc42730637b9313/app/parsers/bulkrax/parser_export_record_set.rb#L174-L183).  I would also consider extracting this logic into a class; because what you are writing in the controller is a pain to test in a controller.

File: hydra/app/jobs/generate_image_thumbs_job.rb [hunk 5]
```diff
@@ -3,42 +3,25 @@ class GenerateImageThumbsJob < ApplicationJob
 
   queue_as :import
 
-  def perform(identifier)
+  def perform(id, download_path)
+    return unless File.exist? download_path
+
     # find record
-    record = Acda.where(identifier: identifier).first
+    record = Acda.find(id)
```
Comment (jeremyf) [inline hydra/app/jobs/generate_image_thumbs_job.rb]: This is a direct Fedora find and is by nature slower.  Is that intentional?

File: hydra/app/jobs/generate_image_thumbs_job.rb [hunk 6]
```diff
@@ -3,42 +3,25 @@ class GenerateImageThumbsJob < ApplicationJob
 
   queue_as :import
 
-  def perform(identifier)
+  def perform(id, download_path)
```
Comment (jeremyf) [inline hydra/app/jobs/generate_image_thumbs_job.rb:6]: Why the change from `id` to `identifier`?  Also in changing the method signature there could be impacts for enqueued jobs that are present between deploys.

Changing the name of the parameter from `id` to `identifier` created lots of little changes in the method.

File: hydra/app/jobs/generate_pdf_thumbs_job.rb [hunk 7]
```diff
@@ -3,44 +3,20 @@ class GeneratePdfThumbsJob < ApplicationJob
 
   queue_as :import
 
-  def perform(identifier)
-    # find record
-    record = Acda.where(identifier: identifier).first
+  def perform(id, download_path)
```
Comment (jeremyf) [inline hydra/app/jobs/generate_pdf_thumbs_job.rb:6]: See the above conversation regarding renaming the parameter.

File: hydra/app/jobs/generate_pdf_thumbs_job.rb [hunk 8]
```diff
@@ -3,44 +3,20 @@ class GeneratePdfThumbsJob < ApplicationJob
...
+    return unless File.exist? download_path
 
+    # find record
+    record = Acda.find(id)
```
Comment (jeremyf) [inline hydra/app/jobs/generate_pdf_thumbs_job.rb]: See the above about shifting from Solr query to Fedora query.  I'm unclear if it is needed.

File: hydra/app/jobs/generate_thumbs_job.rb [hunk 9]
```diff
@@ -1,9 +1,9 @@
 class GenerateThumbsJob < ApplicationJob
   queue_as :import
 
-  def perform(identifier)
+  def perform(id)
     # find record
-    record = Acda.where(identifier: identifier).first
+    record = Acda.find id
```
Comment (jeremyf) [inline hydra/app/jobs/generate_thumbs_job.rb]: See the above about method parameter name and changing from Solr query to Fedora.

File: hydra/app/views/catalog/_document.html.erb [hunk 10]
```diff
@@ -1,25 +1,25 @@
...
    else
 	 @page_title = "Browsing #{application_name}"
-   end 
+   end
 %>
 
 <% # container for a single doc -%>
 <div class="document <%= render_document_class document %>" itemscope itemtype="<%= document.itemtype %>">
 	<div class="document-wrap">
-        <% identifier = document[:id] %>
```
Comment (jeremyf) [inline hydra/app/views/catalog/_document.html.erb:11]: See renaming variable; we have a case where we've renamed identifier to id and id to identifier.  That adds a lot of chatter to the PR.

File: hydra/app/views/catalog/_pdf.html.erb [hunk 11]
```diff
@@ -1,19 +1,19 @@
...
 <div class="row">
-  <div class="col-md-4">      
+  <div class="col-md-4">
       <a class="document-image-full" href="<%= preview %>">
-        <% if Acda.where(identifier: identifier).first&.thumbnail_file.nil? %>
+        <% if @thumbnail_file.nil? %>
           <span class="fa fa-file-pdf-o fa-5x" aria-hidden="true"></span>
         <% else %>
-          <%= image_tag("/image/#{identifier}", title:"#{@document[:title_ssi]}", alt:description, class:"full-size-responsive") %> 
+          <%= image_tag("/image/#{id}", title:"#{@document[:title_ssi]}", alt:description, class:"full-size-responsive") %>
```
Comment (jeremyf) [inline hydra/app/views/catalog/_pdf.html.erb:14]: Again curious about the rename.

File: hydra/app/controllers/catalog_controller.rb (hunk 3 above)
Comment (kirkkwang) [reply hydra/app/controllers/catalog_controller.rb]: code comment is not ours, i'm not removing it just in case they still want to reference it

File: hydra/app/jobs/generate_image_thumbs_job.rb (hunk 5 above)
Comment (kirkkwang) [reply hydra/app/jobs/generate_image_thumbs_job.rb]: i did not know it was slower, so instead do a `Acda.where(id: id).first`?

File: hydra/app/jobs/generate_image_thumbs_job.rb (hunk 6 above)
Comment (kirkkwang) [reply hydra/app/jobs/generate_image_thumbs_job.rb:6]: i thought changing would be better because i'm not passing in an identifier, i'm passing in an id, should i keep calling it an identifier even though it's an id?

File: hydra/app/jobs/generate_pdf_thumbs_job.rb (hunk 7 above)
Comment (kirkkwang) [reply hydra/app/jobs/generate_pdf_thumbs_job.rb:6]: see above

File: hydra/app/views/catalog/_pdf.html.erb (hunk 11 above)
Comment (kirkkwang) [reply hydra/app/views/catalog/_pdf.html.erb:14]: i'm curious why not rename it? if it's not an identifier and it is an id, why keep calling it an identifier?

File: hydra/app/jobs/generate_image_thumbs_job.rb (hunk 6 above)
Comment (kirkkwang) [reply hydra/app/jobs/generate_image_thumbs_job.rb:6]: i added a comment in the model where the thumbnail jobs get called, that should clear up the confusion that this is an ActiveFedora id and not an identifier

File: hydra/app/jobs/generate_thumbs_job.rb (hunk 9 above)
Comment (kirkkwang) [reply hydra/app/jobs/generate_thumbs_job.rb]: changed to a where instead of a find

File: hydra/app/views/catalog/_document.html.erb (hunk 10 above)
Comment (kirkkwang) [reply hydra/app/views/catalog/_document.html.erb:11]: acknowledged that it adds chatter, unfortunately that's just how it will be unless we want to keep it misnamed

File: hydra/app/controllers/catalog_controller.rb (hunk 4 above)
Comment (kirkkwang) [reply hydra/app/controllers/catalog_controller.rb]: Good call!  This lead me to an entirely different solution.  Ultimately we just wanted to check if the image_file (or thumbnail_file) exists on the record.  So now i'm indexing that so we only need to interact with the solr doc

```rb
solr_doc['has_image_file_bsi'] = object.image_file.present?
solr_doc['has_thumbnail_file_bsi'] = object.thumbnail_file.present?
```

File: hydra/app/jobs/generate_thumbs_job.rb [hunk 12]
```diff
@@ -1,27 +1,54 @@
...
+      return unset_image_and_thumbnail!(record)
     end
   end
 
+  private
+
+    def unset_image_and_thumbnail!(record)
+      record.thumbnail_file = nil
+      record.image_file = nil
+      record.update_index
```
Comment (jeremyf) [inline hydra/app/jobs/generate_thumbs_job.rb:55]: To be clear this will not update the underlying Fedora object.  Is that intended?

Comment (kirkkwang) [reply hydra/app/jobs/generate_thumbs_job.rb:55]: I'll say that when i didn't have it, it didn't update and when i added it it did, so without it i'm getting undesired behavior

Comment (kirkkwang) [reply hydra/app/jobs/generate_thumbs_job.rb:55]: also i'm updating the solr index, not the fedora object

File: hydra/app/models/acda.rb [hunk 13]
```diff
@@ -16,7 +16,10 @@ def saved_change_to_preview?
 
   def generate_thumbnail
     # queue job to generate thumbnail
-    GenerateThumbsJob.perform_later(identifier)
+    # we are passing in the Fedora object's id instead of identifier
+    # because the identifier can have special characters which will be
+    # problematic in file and folder creation
+    GenerateThumbsJob.perform_later(id)
```
Comment (jeremyf) [inline hydra/app/models/acda.rb:19]: I was expecting the documentation to be added to `GenerateThumbsJob#perform` to explain the parameter.  The receiver (e.g. `#perform` is a good place for that documentation as it conveys that information to LSP help text and via yard).  This inline comment is lost to any supporting development tools (except at this particular location)

**Review by jeremyf — APPROVED** [review]
I left non-blocking comments

File: hydra/app/models/acda.rb (hunk 13 above)
Comment (kirkkwang) [reply hydra/app/models/acda.rb:19]: can do, i'll move it there instead
